Source/Utils/Average.cpp

```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
+    sf[2] = 0;
+#elif (AMREX_SPACEDIM == 3)
+    sf[2] = stag_fp[2];
+#endif
+
+    sc[0] = stag_cp[0];
+    sc[1] = stag_cp[1];
+#if   (AMREX_SPACEDIM == 2)
+    sc[2] = 0;
+#elif (AMREX_SPACEDIM == 3)
+    sc[2] = stag_cp[2];
+#endif
+
+    cr[0] = ratio[0];
+    cr[1] = ratio[1];
+#if   (AMREX_SPACEDIM == 2)
+    cr[2] = 1;
+#elif (AMREX_SPACEDIM == 3)
+    cr[2] = ratio[2];
+#endif
+
+#ifdef _OPENMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+    // Loop over boxes (or tiles if not on GPU)
+    for (MFIter mfi( mf_cp, TilingIfNotGPU() ); mfi.isValid(); ++mfi)
+    {
+        // Tiles defined at the coarse level
+        const Box& bx = mfi.tilebox();
+        Array4<Real> const& mf_cp_arr = mf_cp.array( mfi );
+        Array4<Real const> const& mf_fp_arr = mf_fp.const_array( mfi );
+        ParallelFor( bx, ncomp,
+                     [=] AMREX_GPU_DEVICE( int i, int j, int k, int n )
+                     {
+                         mf_cp_arr(i,j,k,n+dcomp) = Average::CoarsenAndInterpolateKernel(
+                             mf_fp_arr, sf, sc, cr, i, j, k, n+scomp );
+                     } );
+    }
+}
+
+void
+Average::CoarsenAndInterpolate ( MultiFab& mf_cp,
+                                 const MultiFab& mf_fp,
+                                 const int dcomp,
+                                 const int scomp,
+                                 const int ncomp,
+                                 const IntVect ratio )
+{
+    BL_PROFILE( "Average::CoarsenAndInterpolate" );
+
+    AMREX_D_TERM( AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[0] ),
+                      "input MultiFab is not coarsenable along x" );,
+                  AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[1] ),
+                      "input MultiFab is not coarsenable along y" );,
+                  AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[2] ),
+                      "input MultiFab is not coarsenable along z" ); );
+
+    // Coarsen fine data
+    BoxArray coarsened_mf_fp_ba = mf_fp.boxArray();
+    coarsened_mf_fp_ba.coarsen( ratio );
+
+    if (coarsened_mf_fp_ba == mf_cp.boxArray() and mf_fp.DistributionMap() == mf_cp.DistributionMap())
+        Average::CoarsenAndInterpolateLoop( mf_cp, mf_fp, dcomp, scomp, ncomp, ratio );
+    else
+    // Copy from component scomp of the fine FArrayBox into component 0 of the coarse
+    // FArrayBox because the coarse FArrayBox is a temporary FArrayBox starting at
+    // component 0 and is not part of the actual coarse MultiFab mf_cp
+    {
+        MultiFab coarsened_mf_fp( coarsened_mf_fp_ba, mf_fp.DistributionMap(), ncomp, 0, MFInfo(), FArrayBoxFactory() );
+        Average::CoarsenAndInterpolateLoop( coarsened_mf_fp, mf_fp, 0, scomp, ncomp, ratio );
+        // Copy starts at component 0 in the temporary source MultiFab coarsend_mf_fp
+        mf_cp.copy( coarsened_mf_fp, 0, dcomp, ncomp );
+    }
```

Comment:
That's neat. Minor suggestion, just about comments:
```suggestion
    {
        // Cannot coarsen directly into a MultiFab with different BoxArray/DistributionMapping.
        // Hence, do it in two steps:

        // Temp MultiFab on coarsened version of mf_fp.boxArray(), with same distribution mapping.
        MultiFab coarsened_mf_fp( coarsened_mf_fp_ba, mf_fp.DistributionMap(), ncomp, 0, MFInfo(), FArrayBoxFactory() );
        // First, do the interpolation from mf_fp to coarsened_mf_fp (start writing into component 0).
        Average::CoarsenAndInterpolateLoop( coarsened_mf_fp, mf_fp, 0, scomp, ncomp, ratio );
        // Second, copy from coarsened_mf_fp to mf_cp (with different BoxArray/DistributionMapping).
        mf_cp.copy( coarsened_mf_fp, 0, dcomp, ncomp );
    }
```